typing/typedecl.ml

```diff
@@ -1426,63 +1441,64 @@ let transl_with_constraint env id row_path orig_decl sdecl =
       type_separability = Types.Separability.default_signature ~arity;
       type_is_newtype = false;
       type_expansion_scope = Btype.lowest_level;
-      type_loc = sdecl.ptype_loc;
+      type_loc = loc;
       type_attributes = sdecl.ptype_attributes;
       type_immediate = Unknown;
       type_unboxed;
       type_uid = Uid.mk ~current_unit:(Env.get_unit_name ());
     }
   in
   begin match row_path with None -> ()
-  | Some p -> set_fixed_row env sdecl.ptype_loc p decl
+  | Some p -> set_fixed_row sig_env loc p new_sig_decl
   end;
-  begin match Ctype.closed_type_decl decl with None -> ()
-  | Some ty -> raise(Error(sdecl.ptype_loc, Unbound_type_var(ty,decl)))
+  begin match Ctype.closed_type_decl new_sig_decl with None -> ()
+  | Some ty -> raise(Error(loc, Unbound_type_var(ty, new_sig_decl)))
   end;
-  let decl = name_recursion sdecl id decl in
+  let new_sig_decl = name_recursion sdecl id new_sig_decl in
   let new_type_variance =
-    try Typedecl_variance.compute_decl
-          env ~check:true decl (Typedecl_variance.variance_of_sdecl sdecl)
+    let required = Typedecl_variance.variance_of_sdecl sdecl in
+    try
+      Typedecl_variance.compute_decl sig_env ~check:true new_sig_decl required
```

Comment:
You are right: since the `type_kind` part comes from the signature, we need to use `sig_env` here.